Add pose sampler of paths in rings

aysecanuenal opened the pull request.

alisterburt commented on src/morphosamplers/samplers/path_samplers/pose_sampler_rings.py:
```diff
@@ -63,22 +63,42 @@ def sample(self, obj: Path) -> PoseSet:
+        # target shape (n_path_samples, n_points_per_ring, 3)
+        # orientations = einops.rearrange(orientations, 'n_path_samples i j -> 1 n_path_samples i j')
+        # circle_xyz = einops.rearrange(circle_xyz, 'n_points_per_ring xyz -> n_points_per_ring 1 xyz 1')
+        circle_xyz = einops.rearrange(circle_xyz, 'n_points_per_ring xyz -> 1 xyz n_points_per_ring')
+        rotated_circle_xyz = orientations @ circle_xyz  # (n_points_per_ring, 3, n_path_samples)
```
It's great that this code it producing the correct output! I think it's important that the code is understandable and the semantics feel unclear to me here.

In my head, we are trying to rotate each point in the circle using the rotation matrix (orientation) at each sample point along the path. When we apply a rotation matrix we [left multiply a column vector](https://en.wikipedia.org/wiki/Rotation_matrix#In_three_dimensions) to apply the linear transformation to the points in the circke. 

This is what was being done in the code you've commented out and is not the way the code you replaced it with is structured... it's possible that the code I wrote is wrong somehow but even if that's the case, I'd like to see your rotations applied by left multiplying column vectors so that it's easy to understand the semantics of what's going on 🙂 

Does this make sense?

alisterburt commented on src/morphosamplers/samplers/path_samplers/pose_sampler_rings.py:
```diff
@@ -63,22 +63,42 @@ def sample(self, obj: Path) -> PoseSet:
         # place points around each path sample
-        final_positions = rotated_circle_xyz + path_sample_positions  # (n_points_per_ring, n_path_samples, 3)
+        path_sample_positions_for_ring = einops.rearrange(path_sample_positions, 'n xyz -> n 1 xyz')
+        final_positions = rotated_circle_xyz + path_sample_positions_for_ring # (n_path_samples, n_points_per_ring, 3)
```
nomenclature is a little off here, maybe `ring_per_path_sample_xyz`?

alisterburt commented on src/morphosamplers/samplers/path_samplers/pose_sampler_rings.py:
```diff
@@ -63,22 +63,42 @@ def sample(self, obj: Path) -> PoseSet:
 
-        ## ayse's job - construct the orientations for each particle
-        # finding vector to corrsponding path sample for each particle
+        repeated_path_sample_positions_for_vectors = einops.repeat(path_sample_positions, 'n xyz -> n b xyz', b=self.n_points_per_ring)
```
Great! You're doing this as part of the logic for constructing particle orientations, right? Should this be moved into the block below to reflect that?

alisterburt commented on src/morphosamplers/samplers/path_samplers/pose_sampler_rings.py:
```diff
@@ -63,22 +63,42 @@ def sample(self, obj: Path) -> PoseSet:
+        # (n_path_samples, n_points_per_ring, 3 ,3)
+        outward_z_vectors = final_positions - repeated_path_sample_positions_for_vectors
+        outward_z_vectors /= np.linalg.norm(outward_z_vectors, axis=2, keepdims=True)
+        repeated_path_z_vectors = einops.repeat(z_vectors, 'n xyz -> n b xyz', b=self.n_points_per_ring)
```
would it be worth renaming `z_vectors` to `path_sample_z_vectors` above for clarity?

alisterburt commented on src/morphosamplers/samplers/path_samplers/pose_sampler_rings.py:
```diff
@@ -63,22 +63,42 @@ def sample(self, obj: Path) -> PoseSet:
         # - x vector perpendicular to both
-        # construct (n_points_per_ring, n_path_samples, 3, 3) orientations
+        # (n_path_samples, n_points_per_ring, 3 ,3)
+        outward_z_vectors = final_positions - repeated_path_sample_positions_for_vectors
```
```suggestion
        outward_z_vectors = final_positions - repeated_path_sample_positions_for_vectors
```

`outward_z_vectors` -> `z_vectors`?

alisterburt commented on src/morphosamplers/samplers/path_samplers/pose_sampler_rings.py:
```diff
@@ -63,22 +63,42 @@ def sample(self, obj: Path) -> PoseSet:
+        repeated_path_z_vectors = einops.repeat(z_vectors, 'n xyz -> n b xyz', b=self.n_points_per_ring)
+        y_vectors = repeated_path_z_vectors # along the path, calculated above
+        y_vectors /= np.linalg.norm(y_vectors, axis=2, keepdims=True)
+        x_vectors = np.cross(outward_z_vectors, repeated_path_z_vectors)
```
```suggestion
        x_vectors = np.cross(z_vectors, y_vectors)
```

I understand what's going on here now but it took me a few times reading - have suggested some renames above that should clarify - there are lots of coordinate systems here and clarity is hard! 😂

alisterburt commented on src/morphosamplers/samplers/path_samplers/pose_sampler_rings.py:
```diff
@@ -63,22 +63,42 @@ def sample(self, obj: Path) -> PoseSet:
+        x_vectors = np.cross(outward_z_vectors, repeated_path_z_vectors)
+        x_vectors /= np.linalg.norm(x_vectors, axis=2, keepdims=True)
+
+        # Make the final_positions and final_orientations suitable for the PoseSet
+        final_positions = einops.rearrange(final_positions, 'b1 b2 xyz -> (b1 b2) xyz')
+        final_orientations = einops.rearrange([x_vectors, y_vectors, outward_z_vectors], 'v b1 b2 xyz -> (b1 b2) xyz v')
```
nice einops!!

alisterburt reviewed (commented):
Amazing work @aysecanuenal ! The code generally reads well and is clear - I've made a few suggestions about names and would love it if you could restructure the application of the path sample rotation (orientation) to the so that it follows standard matrix multiplication semantics. I've described this in more detail below and am happy to work through this with you if you struggle with this restructuring. 

Well done! This stuff is really not trivial and you seem to have taken it in your stride 🙂  super proud of you

aysecanuenal commented:
Hi @alisterburt thank you very much for the comments. Yes, they all make a lot of sense, I think I got what you meant with the multiplication and semantics after going back and looking at it in detail. I changed it back to the correct version and modified the rest according to that

alisterburt commented:
Wonderful!!!
